src/java/org/apache/cassandra/db/compaction/CompactionTask.java

```diff
@@ -205,24 +209,31 @@ public boolean apply(SSTableReader sstable)
                             }
                         }
 
-                        // don't replace old sstables yet, as we need to mark the compaction finished in the system table
+                        readyToFinish = true;
                         newSStables = writer.finish();
                     } catch (Exception e) {
-                        throw new CompactionException(taskIdLoggerMsg, ssTableLoggerMsg.toString(), e);
-                    }
-                    finally
-                    {
-                        Directories.removeExpectedSpaceUsedByCompaction(expectedWriteSize, CONSIDER_CONCURRENT_COMPACTIONS);
-                        // point of no return -- the new sstables are live on disk; next we'll start deleting the old ones
-                        // (in replaceCompactedSSTables)
-                        if (taskId != null)
-                            SystemKeyspace.finishCompaction(taskId);
-
-                        if (collector != null)
-                            collector.finishCompaction(ci);
+                        CompactionException exception = new CompactionException(taskIdLoggerMsg, ssTableLoggerMsg.toString(), e);
+                        if (readyToFinish && e.getSuppressed() != null && e.getSuppressed().length != 0)
+                        {
+                            abortFailed = true;
+                            logger.warn("CompactionAwareWriter failed to close correctly for {}/{}. This compaction won't be removed from " +
+                                            "system.compactions_in_progress to ensure sstable cleanup on startup proceeds correctly in case some " +
+                                            "compaction-product sstables are marked final while others remain tmp",
+                                    cfs.keyspace.getName(), cfs.name, e);
+                        }
+                        throw exception;
                     }
                 }
             }
+            finally
+            {
+                Directories.removeExpectedSpaceUsedByCompaction(expectedWriteSize, CONSIDER_CONCURRENT_COMPACTIONS);
+                if (taskId != null && (!abortFailed))
+                    SystemKeyspace.finishCompaction(taskId);
+
+                if (collector != null && ci != null)
+                    collector.finishCompaction(ci);
+            }
```

Comment:
the abortFailed/readyToFinish flags and checking for suppressed exceptions are working against the intention of the try-with-resources statement, which is to always run all close() method regardless of if the previous ones failed, and suppress the failures from closing https://docs.oracle.com/javase/8/docs/technotes/guides/language/try-with-resources.html, whereas here we explicitly want to skip the SystemKeyspace.finishCompaction() if closing failed. approving the PR for now, but i think this critical section would be less of a workaround + more idiomatic/readable like this:
```
            boolean abortFailed = false;
            Optional<UUID> taskId = Optional.empty();
            String taskIdLoggerMsg = UUIDGen.getTimeUUID().toString();
            List<SSTableReader> newSStables;
            try (Refs<SSTableReader> refs = Refs.ref(actuallyCompact);
                 AbstractCompactionStrategy.ScannerList scanners = strategy.getScanners(actuallyCompact))
            {
                if (!offline) {
                    taskId = Optional.of(SystemKeyspace.startCompaction(cfs, transaction.originals()));
                    taskIdLoggerMsg = taskId.toString();
                }
                logger.debug("Compacting ({}) {}", taskIdLoggerMsg, ssTableLoggerMsg);
                AbstractCompactionIterable ci = new CompactionIterable(compactionType, scanners.scanners, controller, sstableFormat, taskId);
                try (CloseableIterator<AbstractCompactedRow> iter = ci.iterator())
                {
                    long lastCheckObsoletion = start;

                    if (!controller.cfs.getCompactionStrategy().isActive)
                        throw new CompactionInterruptedException(ci.getCompactionInfo());

                    CompactionAwareWriter writer = getCompactionAwareWriter(cfs, transaction, actuallyCompact);
                    try (CompactionExecutorStatsTracker statsTracker = new CompactionExecutorStatsTracker(ci))
                    {
                        estimatedKeys = writer.estimatedKeys();
                        while (iter.hasNext())
                        {
                            if (ci.isStopRequested())
                                throw new CompactionInterruptedException(ci.getCompactionInfo());

                            try (AbstractCompactedRow row = iter.next())
                            {
                                if (writer.append(row))
                                    totalKeysWritten++;

                                if (System.nanoTime() - lastCheckObsoletion > TimeUnit.MINUTES.toNanos(1L))
                                {
                                    controller.maybeRefreshOverlaps();
                                    lastCheckObsoletion = System.nanoTime();
                                }
                            }
                        }

                        newSStables = writer.finish();
                    } finally {
                        writer.close();
                        taskId.ifPresent(SystemKeyspace::finishCompaction);
                    }
                }
            } catch (CompactionInterruptedException e) {
                throw e;
            } catch (Exception e) {
                throw new CompactionException(taskIdLoggerMsg, ssTableLoggerMsg.toString(), e);
            } finally {
                Directories.removeExpectedSpaceUsedByCompaction(expectedWriteSize, CONSIDER_CONCURRENT_COMPACTIONS);
            }

    // ... later inside the toplevel class ...
    private CompactionExecutorStatsTracker implements AutoCloseable {
        private final AbstractCompactionIterable ci;

        public ComapctionExecutorStatsTracker(AbstractCompactionIterable ci) {
            this.ci = ci;
            if (collector != null) {
                collector.beginCompaction(this.ci);
            }
        }
        
        @Override
            public void close() {
            if (collector != null) {
                collector.finishCompaction(ci);
            }
        }
    }
```

Reply:
I will follow up with this one!